pkg/framework/pkcs12.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2020 Red Hat, Inc. and/or its affiliates
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package framework
+
+import (
+	"crypto/rand"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	v1 "k8s.io/api/core/v1"
+	"software.sslmate.com/src/go-pkcs12"
+	"strings"
+)
+
+const (
+	certificateBlock    = "CERTIFICATE"
+	certificateEndBlock = "-----END CERTIFICATE-----"
+)
+
+// CreatePKCS12TrustStoreFromSecret creates a PCKS12 with certificates inside a Kubernetes secret.
+// The `keys` parameter must contain public certificates only.
+func CreatePKCS12TrustStoreFromSecret(secret *v1.Secret, password string, keys ...string) ([]byte, error) {
+	if len(keys) == 0 {
+		return nil, errors.New("at least one key is required")
+	}
+	var crts []string
+	for _, key := range keys {
+		cert := string(secret.Data[key])
+		crts = append(crts, splitCertificates(cert, 0, crts)...)
+	}
+	if len(password) == 0 {
+		password = pkcs12.DefaultPassword
+	}
+	var certs []*x509.Certificate
+	for _, crt := range crts {
+		block, _ := pem.Decode([]byte(crt))
+		if block == nil || block.Type != certificateBlock {
+			continue
+		}
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return nil, err
+		}
+		certs = append(certs, cert)
+	}
+
+	pfx, err := pkcs12.EncodeTrustStore(rand.Reader, certs, password)
+	return pfx, err
+}
+
+func splitCertificates(certificate string, index int, certificates []string) []string {
```

Comment:
What is the purpose of this function?

Reply:
To split the certificates contained in a single `crt` file, example:

```
-----BEGIN CERTIFICATE-----
(...)
-----END CERTIFICATE-----
-----BEGIN CERTIFICATE-----
(...)
-----END CERTIFICATE-----
```

A single `crt` can contain many certificates, `pem` and `x509` packages only parse the first one. We need the entire chain in the Trust Store.

Reply:
I see, thanks